### PR TITLE
[GSoC] Added Sampling of Stochastic Process(DTMC)

### LINKED ISSUE #19454: [GSoC Discussion] Sampling from Stochatic Processes
After completing the sampling from the external libraries, we can think of sampling from the stochastic processes. Something similar to the sample path that is followed by the stochastic process, we can use the same for random sampling at any given time. One of the examples, I found is [http://theanalysisofdata.com/probability/7_2.html](http://theanalysisofdata.com/probability/7_2.html)

### PR DESCRIPTION
**Pull request by Smit-create**
Added sampling of Stochastic Process
#### References to other Issues or PRs

FIxes #19454

#### Release Notes

* stats
  * Added `sample_stochastic` for sampling from stochastic processes.

ping @czgdp1807 @Upabjojr

### CODE REVIEW

**Comment by Smit-create** [comment]
@czgdp1807  If this implementation looks good then I will add docstrings of `sample_stochastic`. Moreover, it works as expected by you. With `max_time=S.Infinity` by default, we can sample for as much time as needed. Also, once this gets finalized, I can add the `sample` method for Counting Process in the open PR itself.

File: sympy/stats/tests/test_stochastic_process.py [hunk 1]
```diff
@@ -108,6 +112,29 @@ def test_DiscreteMarkovChain():
     assert variance(X[1], Eq(X[0], 1)) == Rational(8, 9)
     raises(ValueError, lambda: E(X[1], Eq(X[2], 1)))
 
+
+def test_sample_stochastic():
```
Comment (czgdp1807) [inline sympy/stats/tests/test_stochastic_process.py:116]: Are these tests deterministic? I cannot find the value of seed being fixed for these tests in the diff.

File: sympy/stats/stochastic_process_types.py [hunk 2]
```diff
@@ -189,6 +190,9 @@ def joint_distribution(self, *args):
     def expectation(self, condition, given_condition):
         raise NotImplementedError("Abstract method for expectation queries.")
 
+    def sample(self, max_time):
+        raise NotImplementedError()
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:194]: Can you pass a message here like, reached an abstract method?

File: sympy/stats/tests/test_stochastic_process.py (hunk 1 above)
Comment (Smit-create) [reply sympy/stats/tests/test_stochastic_process.py:116]: They are random! Do you mean using sampling seeds to make them produce same result always?

File: sympy/stats/stochastic_process_types.py [hunk 3]
```diff
@@ -727,6 +731,34 @@ def limiting_distribution(self):
...
+        """
+        if not isinstance(self.transition_probabilities, (Matrix, ImmutableMatrix)):
+            raise ValueError("Transition Matrix must be provided for sampling")
+        Tlist = self.transition_probabilities.tolist()
+        samps = [random.choice(list(self.state_space))]
+        yield samps[0]
+        time = 1
+        while time < max_time:
+            dist = Tlist[samps[time - 1]]
+            density = {list(self.state_space)[i]: dist[i] for i in range(len(dist))}
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:757]: Can you move this outside while loop?

Comment (czgdp1807) [reply sympy/stats/stochastic_process_types.py:757]: I see that, density for all the states can be computed before starting the sampling loop.

File: sympy/stats/tests/test_stochastic_process.py (hunk 1 above)
Comment (czgdp1807) [reply sympy/stats/tests/test_stochastic_process.py:116]: Yes, tests should be deterministic. They can be done adding a line like,
```suggestion
def test_sample_stochastic():
    random.seed(0)
```

**Comment by czgdp1807** [comment]
> Also, once this gets finalized, I can add the `sample` method for Counting Process in the open PR itself.

Make a new PR after this gets merged.

**Comment by czgdp1807** [comment]
LGTM. Will merge tomorrow. @Upabjojr @jmig5776 Any thoughts?

**Comment by Upabjojr** [comment]
Is the max_time really needed in `sample_stochastic`?

The name `sample_stochastic` is a bit strange... it's meant to sample a stochastic process? Shouldn't it be named `sample_stochastic_process`? The word _stochastic_ is just a synonym of _random_ in the end.

**Comment by Upabjojr** [comment]
My doubt about the max time variable is that the stochastic process sampled may be different... a maximum time may introduce a sampling bias.

**Comment by Upabjojr** [comment]
Or better said... are you sure that estimators applied to a series generated with a maximum time will return the same values as estimators on the infinitely long series?

**Comment by Smit-create** [comment]
> Shouldn't it be named `sample_stochastic_process`?

Yes, I will change it!

> are you sure that estimators applied to a series generated with a maximum time will return the same values as estimators on the infinitely long series?

I think that this should not affect the sampling, as max_time comes only in the picture when we need to stop the iteration... Still I agree that we can remove this as an argument and use `sample_stochastic_process` with the help of `next` until the user wants it to get a sample. This also seems to work good  enough.

**Comment by czgdp1807** [comment]
`max_time` can be removed. User can do, `next` until he/she wants.

**Pull request merged**